src/cose/sign1.rs

```diff
@@ -0,0 +1,510 @@
+///     .payload(b"This is the content.".to_vec());
+/// let prepared = PreparedCoseSign1::new(builder, None, None, true).unwrap();
+/// let signature_payload = prepared.signature_payload();
+/// let signature = sign::<SigningKey, Signature>(signature_payload, &signer).unwrap();
+/// let cose_sign1 = prepared.finalize(signature);
+///
+/// fn sign<S, Sig>(signature_payload: &[u8], s: &S) -> anyhow::Result<Vec<u8>>
+/// where
+///     S: Signer<Sig> + SignatureAlgorithm,
+///     Sig: SignatureEncoding,
+/// {
+///     Ok(s.try_sign(signature_payload)
+///         .map_err(Error::Signing)?
+///         .to_vec())
+/// }
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PreparedCoseSign1 {
+    tagged: bool,
+    cose_sign1: CoseSign1,
+    signature_payload: Vec<u8>,
+}
+
+/// COSE_Sign1 implementation.
+#[derive(Clone, Debug)]
+pub struct CoseSign1 {
+    tagged: bool,
+    pub inner: coset::CoseSign1,
+}
+
+/// Errors that can occur when building, signing or verifying a COSE_Sign1.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("the COSE_Sign1 has an attached payload but an detached payload was provided")]
+    DoublePayload,
+    #[error("the COSE_Sign1 has a detached payload which was not provided")]
+    NoPayload,
+    #[error("signature did not match the structure expected by the verifier: {0}")]
+    MalformedSignature(signature::Error),
+    #[error("signature is already present")]
+    AlreadySigned,
+    #[error("error occurred when signing COSE_Sign1: {0}")]
+    Signing(signature::Error),
+    #[error("unable to set ClaimsSet: {0}")]
+    UnableToDeserializeIntoClaimsSet(CoseError),
+}
+
+/// Result with error type: [`Error`].
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// Result for verification of a COSE_Sign1.
+#[derive(Debug)]
+pub enum VerificationResult {
+    Success,
+    Failure(String),
+    Error(Error),
+}
+
+impl VerificationResult {
+    /// Result of verification.
+    ///
+    /// False implies the signature is inauthentic or the verification algorithm encountered an
+    /// error.
+    pub fn success(&self) -> bool {
+        matches!(self, VerificationResult::Success)
+    }
+
+    /// Translate to a std::result::Result.
+    ///
+    /// Converts failure reasons and errors into a String.
+    pub fn to_result(self) -> Result<(), String> {
+        match self {
+            VerificationResult::Success => Ok(()),
+            VerificationResult::Failure(reason) => Err(reason),
+            VerificationResult::Error(e) => Err(format!("{}", e)),
+        }
+    }
+
+    /// Retrieve the error if the verification algorithm encountered an error.
+    pub fn to_error(self) -> Option<Error> {
+        match self {
+            VerificationResult::Error(e) => Some(e),
+            _ => None,
+        }
+    }
+}
+
+impl PreparedCoseSign1 {
+    pub fn new(
+        builder: coset::CoseSign1Builder,
+        detached_payload: Option<Vec<u8>>,
+        aad: Option<Vec<u8>>,
```

Comment:
Avoid passing by move `Vec<u8>` if you can pass `&[u8]`. Stuff like `aad` may be reused outside of this function so its good to avoid clones if possible.